src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -3171,10 +3188,13 @@ protected virtual void MarkInstruction (Instruction instruction, MethodDefinitio
 					Annotations.MarkRelevantToVariantCasting (operand.Resolve ());
 					break;
 				case Code.Isinst:
+					if (operand is TypeSpecification || operand is GenericParameter)
+						break;
+
 					if (Annotations.GetAction (method.DeclaringType.Module.Assembly) != AssemblyAction.Link)
 						break;
 
-					if (operand is TypeSpecification || operand is GenericParameter)
+					if (!_context.IsOptimizationEnabled (CodeOptimizations.UnusedTypeChecks, method.DeclaringType))
```

Comment:
These two statements both need the assembly of the method - extract to local variable and reuse.

Reply:
Fixed